File: javascript/13-chatterbox/src/inventory_item.js
class InventoryItem {
  constructor(id, name, batchNumber, cacheTimestamp, rowId, quantity) {
    this.id = id;
    this.name = name;
    this.batchNumber = batchNumber;
    this.cacheTimestamp = cacheTimestamp;
    this.rowId = rowId;
    this.quantity = quantity;
  }

  get_id() {
    return this.id;
  }

  set_id(val) {
    this.id = val;
  }

  get_name() {
    return this.name;
  }

  set_name(val) {
    this.name = val;
  }

  get_batch_number() {
    return this.batchNumber;
  }

  set_batch_number(val) {
    this.batchNumber = val;
  }

  get_cache_timestamp() {
    return this.cacheTimestamp;
  }

  set_cache_timestamp(val) {
    this.cacheTimestamp = val;
  }

  get_row_id() {
    return this.rowId;
  }

  set_row_id(val) {
    this.rowId = val;
  }

  get_quantity() {
    return this.quantity;
  }

  set_quantity(val) {
    this.quantity = val;
  }

  reserve(units) {
    if (units <= 0) {
      return { status: 'rejected', reason: 'invalid_quantity', remaining: this.quantity };
    }
    if (units > this.quantity) {
      return { status: 'backorder', reserved: 0, remaining: this.quantity };
    }

    this.quantity -= units;
    return { status: 'reserved', reserved: units, remaining: this.quantity, sku: `${this.id}-${this.batchNumber}` };
  }

  receive_stock(units) {
    if (units <= 0) {
      return this.quantity;
    }

    this.quantity += units;
    return this.quantity;
  }

  public_snapshot() {
    return {
      id: this.id,
      name: this.name,
      batch_number: this.batchNumber,
      quantity: this.quantity,
      stock_status: this.stock_status(),
    };
  }

  stock_status() {
    if (this.quantity === 0) {
      return 'out';
    }
    if (this.quantity < 5) {
      return 'low';
    }
    return 'available';
  }
}

export { InventoryItem };

File: golang/13-chatterbox/inventory_item.go
package kata

import "fmt"

type InventoryItem struct {
	ID             int
	Name           string
	BatchNumber    string
	CacheTimestamp int
	RowID          int
	Quantity       int
}

func NewInventoryItem(id int, name string, batchNumber string, cacheTimestamp int, rowID int, quantity int) *InventoryItem {
	return &InventoryItem{ID: id, Name: name, BatchNumber: batchNumber, CacheTimestamp: cacheTimestamp, RowID: rowID, Quantity: quantity}
}

func (i *InventoryItem) get_id() int              { return i.ID }
func (i *InventoryItem) get_name() string         { return i.Name }
func (i *InventoryItem) get_batch_number() string { return i.BatchNumber }
func (i *InventoryItem) get_cache_timestamp() int { return i.CacheTimestamp }
func (i *InventoryItem) get_row_id() int          { return i.RowID }
func (i *InventoryItem) get_quantity() int        { return i.Quantity }

func (i *InventoryItem) reserve(units int) map[string]interface{} {
	if units <= 0 {
		return map[string]interface{}{"status": "rejected", "reason": "invalid_quantity", "remaining": i.Quantity}
	}
	if units > i.Quantity {
		return map[string]interface{}{"status": "backorder", "reserved": 0, "remaining": i.Quantity}
	}

	i.Quantity -= units
	return map[string]interface{}{
		"status":    "reserved",
		"reserved":  units,
		"remaining": i.Quantity,
		"sku":       fmt.Sprintf("%d-%s", i.ID, i.BatchNumber),
	}
}

func (i *InventoryItem) receive_stock(units int) int {
	if units <= 0 {
		return i.Quantity
	}

	i.Quantity += units
	return i.Quantity
}

func (i *InventoryItem) public_snapshot() map[string]interface{} {
	return map[string]interface{}{
		"id":           i.ID,
		"name":         i.Name,
		"batch_number": i.BatchNumber,
		"quantity":     i.Quantity,
		"stock_status": i.stock_status(),
	}
}

func (i *InventoryItem) stock_status() string {
	if i.Quantity == 0 {
		return "out"
	}
	if i.Quantity < 5 {
		return "low"
	}
	return "available"
}
